### PR TITLE
[Sparse conv] Implement implicit gemm algo for SubmConv3D

### PR DESCRIPTION
**Pull request by Wong4j**
### PR Category
Performance Optimization

### PR Types
Performance

### Description

Support implicit GEMM algorithm for SubmConv3D.

Usage:
- nn.functional.subm_conv3d_igemm
```python
# 3D
y = paddle.sparse.nn.functional.subm_conv3d(x, weight, key='key1', padding=[0, 1, 1])   # original
y = paddle.sparse.nn.functional.subm_conv3d_igemm(x, weight, key='key2', padding=[0, 1, 1])   # use implicit gemm

# 2D
y = paddle.sparse.nn.functional.subm_conv2d(x, weight, key='key1', padding=[1, 1])   # original
y = paddle.sparse.nn.functional.subm_conv2d_igemm(x, weight, key='key2', padding=[1, 1])   # use implicit gemm
```
- nn.SubmConv3D
```python
# 3D
paddle.nn.SubmConv3D(32, 32, kernel_size=[1, 3, 3], key="key1")  # original
paddle.nn.SubmConv3D(32, 32, kernel_size=[1, 3, 3], key="key2", backend="igemm")  # use implicit gemm

# 2D
paddle.nn.SubmConv2D(32, 32, kernel_size=[3, 3], key="key1")  # original
paddle.nn.SubmConv2D(32, 32, kernel_size=[3, 3], key="key2", backend="igemm")  # use implicit gemm
```

Perf:
GPU: 3080
Prec: FP16
case: single SubmConv
nnz=214202  dense_shape=[1, 1, 4608, 4608, 32]  kernel_size=[1, 3, 3] stride=1  in_channel=out_channel=32
(This perf numbers do not include the overhead of hashmap/rulebook creation, which I assume has been cached.)
| --- |  cutlass |   igemm|
|----| ------ | ----- |
| time(us)| 703 | 194|

Note: 
- Implicit gemm only supports forward now.
- I have only verified the correctness for SubmConv, so I'm asserting `subm==True and stride==1 and dilation==1` in the code.
- The cuda kernels are modified based on the [torchsparse](https://github.com/mit-han-lab/torchsparse)'s implementation.
- ~~The input must be 3D (NDHWC), and kernel must has dims=3. For 2D case, please insert zeros to to the D dimention of indices and set `kernel sizes` to (1, 3, 3).~~
- The input can be 3D (NDHWC) or 2D (NHWC)

### CODE REVIEW

File: paddle/phi/infermeta/sparse/binary.cc [hunk 1]
```diff
@@ -121,6 +121,43 @@ void Conv3dInferMeta(const MetaTensor& x,
   counter->set_dims({1});
 }
 
+void Conv3dImplicitGemmInferMeta(const MetaTensor& x,
```
Comment (Wangzheee) [inline paddle/phi/infermeta/sparse/binary.cc:124]: ci的代码没有覆盖到这个OP，可以针对这个OP增加单测

File: paddle/phi/core/kmap_cache.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+struct KmapCache {
+  DenseTensor* out_in_map = nullptr;
+  DenseTensor* coords = nullptr;
+  DenseTensor* hashmap_keys = nullptr;
+  DenseTensor* hashmap_values = nullptr;
+  // std::vector<int>* spatial_range;
+
+  // destructor
+  ~KmapCache() {
```
Comment (Wangzheee) [inline paddle/phi/core/kmap_cache.h:30]: 单测中没有执行这个析构，可以增加一下

Comment (Wong4j) [reply paddle/phi/core/kmap_cache.h:30]: 我增加了单测，本地跑单测会跑到这个析构，但是CI仍然显示没有跑到。

File: paddle/phi/infermeta/sparse/binary.cc (hunk 1 above)
Comment (Wong4j) [reply paddle/phi/infermeta/sparse/binary.cc:124]: Done

**Review by tianshuo78520a — APPROVED** [review]

File: paddle/phi/kernels/sparse/gpu/conv_kernel_impl.cuh [hunk 3]
```diff
@@ -0,0 +1,1273 @@
+#include <cuda_fp16.h>
```
Comment (qingqing01) [inline paddle/phi/kernels/sparse/gpu/conv_kernel_impl.cuh:1]: 文件开头加下 copyright 吧

File: python/paddle/sparse/nn/layer/conv.py [hunk 4]
```diff
@@ -624,6 +672,7 @@ def __init__(
         weight_attr=None,
         bias_attr=None,
         data_format="NDHWC",
+        backend=None,
```
Comment (qingqing01) [inline python/paddle/sparse/nn/layer/conv.py:675]: 对用户暴露的接口需增加注释

**Review by qingqing01 — APPROVED** [review]
后续需要更新中文文档

**Review by Aurelius84 — APPROVED** [review]

File: python/paddle/sparse/nn/layer/conv.py [hunk 5]
```diff
@@ -40,6 +40,7 @@ def __init__(
...
```
Comment (jzhang533) [inline python/paddle/sparse/nn/layer/conv.py:43]: the newly introduced arg `backend` should be documented in the docstring.

File: python/paddle/sparse/nn/layer/conv.py [hunk 6]
```diff
@@ -53,11 +54,16 @@ def __init__(
...
         self._subm = subm
         self._key = key
+        self._backend = backend
 
         assert (
             padding_mode == 'zeros'
         ), "Currently, only support padding_mode='zeros'"
         assert groups == 1, "Currently, only support groups=1"
+        assert backend in [
+            None,
```
Comment (jzhang533) [inline python/paddle/sparse/nn/layer/conv.py:64]: do we need a short string for the original implementation, e.g. : native or cutlass ?
None is not clear for user.

Comment (ming1753) [reply python/paddle/sparse/nn/layer/conv.py:64]: sp_conv框架内原本有两种实现，native和cutlass(部分就绪未编译)，这里提供了新的实现igemm，现在backend为None时走native，指定igemm后走目前性能最好的igemm，cutlass版还需另一个PR合入。在cutlass版就绪后会对这里改动和增加详细注释。

Comment (jzhang533) [reply python/paddle/sparse/nn/layer/conv.py:64]: Let's wait for PR author's opinion.

Comment (Wong4j) [reply python/paddle/sparse/nn/layer/conv.py:64]: 这个igemm是针对百度业务需求加入的特定优化，支持的case比较有限(并没有开发完)，目前你们内部业务反馈符合预期。后续进一步开发完善，由百度的同学接手。
这里只是一个暂时的设计，后面百度的同学可以整合其他backend，比如cutlass，再修改这个选项。

**Review by jzhang533 — APPROVED** [review]
LGTM

**Review by Wangzheee — APPROVED** [review]

**Pull request merged**